File: leetcode/221_maximal_square.py
class Solution(object):

    def get_tl(self, matrix, y, x):
        result = []
        for r, c in [(y-1, x), (y, x-1), (y-1, x-1)]:
            if r >= 0 and c >= 0:
                curr = matrix[r][c]
                result.append(curr)
            else:
                result.append(0)
        return result

    def maximalSquare(self, matrix):
        result = 0
        for y, row in enumerate(matrix):
            for x, num in enumerate(row):
                matrix[y][x] = int(num)
        for y, row in enumerate(matrix):
            for x, num in enumerate(row):
                if not num:
                    continue
                curr = num
                top, left, top_left = self.get_tl(matrix, y, x)
                curr = min(top, left, top_left) + 1
                matrix[y][x] = curr
                result = max(result, curr)

File: leetcode/java/221MaximalSquare.java
/**
 * T: O(M*N) S: O(M*N)
 *
 * <p>On each step we check what the maximum square that can be made of the left, top, and top left
 * of the current cell. If on all of them we can make a square of size S, we can make a square of
 * size S+1 on the current cell.
 */
class Solution {
  public int maximalSquare(char[][] matrix) {
    int rows = matrix.length;
    int cols = rows == 0 ? 0 : matrix[0].length;
    int[][] dp = new int[rows + 1][cols + 1];
    int maxSide = 0;
    for (int r = 1; r <= rows; r++) {
      for (int c = 1; c <= cols; c++) {
        if (matrix[r - 1][c - 1] == '1') {
          dp[r][c] = Math.min(Math.min(dp[r - 1][c], dp[r][c - 1]), dp[r - 1][c - 1]) + 1;
          maxSide = Math.max(maxSide, dp[r][c]);
        }
      }
    }
    return maxSide * maxSide;
  }
}
